File: typescript-examples/playwright-basics-ts/api/05-scrape-list.ts
import { BrowserContext, Page } from "playwright";

/**
 * Scrape List
 *
 * Demonstrates extracting data from a list of elements
 * using count() and nth() to iterate.
 */

interface Params {
  maxItems?: number;
}

interface Product {
  title: string | null;
  price: string | null;
  rating: string | null;
  inStock: boolean;
}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
) {
  await page.goto("https://books.toscrape.com/");
  await page.waitForLoadState("networkidle");

  const products: Product[] = [];
  const productCards = page.locator(".product_pod");
  const totalCount = await productCards.count();
  const maxItems = params.maxItems ?? totalCount;

  for (let i = 0; i < Math.min(totalCount, maxItems); i++) {
    const card = productCards.nth(i);

    // Extract title from the link's title attribute
    const title = await card.locator("h3 a").getAttribute("title");

    // Extract price
    const price = await card.locator(".price_color").textContent();

    // Extract rating from class (e.g., "star-rating Three")
    const ratingClass = await card.locator(".star-rating").getAttribute("class");
    const rating = ratingClass?.split(" ").pop() ?? null;

    // Check if in stock
    const inStock = await card.locator(".instock").isVisible();

    products.push({
      title,
      price: price?.trim() ?? null,
      rating,
      inStock,
    });
  }

  return {
    totalAvailable: totalCount,
    scraped: products.length,
    products,
  };
}

File: python-examples/playwright-basics/api/05-scrape-list.py
"""
Scrape List

Demonstrates extracting multiple items from a page:
- Using count() to get number of elements
- Using nth() to access specific elements in a loop
"""

from typing import TypedDict

from playwright.async_api import Page


class Params(TypedDict):
    limit: int


async def automation(page: Page, params: Params | None = None, **_kwargs):
    limit = params.get("limit", 10) if params else 10

    await page.goto("https://books.toscrape.com/")
    await page.wait_for_load_state("networkidle")

    products = page.locator(".product_pod")
    total_count = await products.count()

    # Limit to requested number
    scrape_count = min(limit, total_count)

    items = []
    for i in range(scrape_count):
        product = products.nth(i)

        # Get title from the link's title attribute
        title = await product.locator("h3 a").get_attribute("title")

        # Get price
        price = await product.locator(".price_color").text_content()

        # Get rating class (e.g., "star-rating Three")
        rating_class = await product.locator(".star-rating").get_attribute("class")
        rating = rating_class.replace("star-rating ", "") if rating_class else None

        # Check if in stock
        in_stock = await product.locator(".instock").count() > 0

        items.append(
            {
                "title": title,
                "price": price,
                "rating": rating,
                "inStock": in_stock,
            }
        )

    return {
        "totalAvailable": total_count,
        "scraped": len(items),
        "products": items,
    }
